packages/SwingSet/src/liveslots/watchedPromises.js

```diff
@@ -0,0 +1,162 @@
+      const watches = watchedPromiseTable.get(vpid);
+      watchedPromiseTable.delete(vpid);
+      for (const watch of watches) {
+        const [watcher, ...args] = watch;
+        const handler = wasFulfilled ? watcher.onFulfilled : watcher.onRejected;
+        if (handler) {
+          handler(value, ...args);
+        }
+      }
+    }
+
+    p.then(
+      res => settle(res, true),
+      rej => settle(rej, false),
+    );
+  }
+
+  function loadWatchedPromiseTable() {
+    const deadPromisesRaw = syscall.vatstoreGet('deadPromises');
+    if (deadPromisesRaw) {
+      const deadPromises = new Set(deadPromisesRaw.split(','));
+      syscall.vatstoreDelete('deadPromises');
+      for (const [vpid, watches] of watchedPromiseTable.entries()) {
+        if (deadPromises.has(vpid)) {
+          watchedPromiseTable.delete(vpid);
+          for (const watch of watches) {
+            const [watcher, ...args] = watch;
+            if (watcher.onRejected) {
+              watcher.onRejected('vat upgraded', ...args);
+            }
+          }
+        } else {
+          const p = convertSlotToVal(vpid);
+          pseudoThen(p, vpid);
+        }
+      }
+    }
+  }
+
+  // defined promise watcher objects indexed by kindHandle
+  let promiseWatcherByKindTable;
+
+  function preparePromiseWatcherTables() {
+    let watcherTableID = syscall.vatstoreGet('watcherTableID');
+    if (watcherTableID) {
+      promiseWatcherByKindTable = convertSlotToVal(watcherTableID);
+    } else {
+      promiseWatcherByKindTable = makeScalarBigMapStore(
+        'promiseWatcherByKind',
+        { durable: true },
+      );
+      watcherTableID = convertValToSlot(promiseWatcherByKindTable);
+      syscall.vatstoreSet('watcherTableID', watcherTableID);
+      // artificially increment the table's refcount so it never gets GC'd
+      vrm.addReachableVref(watcherTableID);
+    }
+
+    let watchedPromiseTableID = syscall.vatstoreGet('watchedPromiseTableID');
+    if (watchedPromiseTableID) {
+      watchedPromiseTable = convertSlotToVal(watchedPromiseTableID);
+    } else {
+      watchedPromiseTable = makeScalarBigMapStore('watchedPromises', {
+        keySchema: M.string(),
+        durable: true,
+      });
+      watchedPromiseTableID = convertValToSlot(watchedPromiseTable);
+      syscall.vatstoreSet('watchedPromiseTableID', watchedPromiseTableID);
+      // artificially increment the table's refcount so it never gets GC'd
+      vrm.addReachableVref(watchedPromiseTableID);
+    }
+  }
+
+  function watchPromise(p, watcher, ...args) {
+    const watcherVref = convertValToSlot(watcher);
+    assert(watcherVref, 'invalid watcher');
+    const { virtual } = parseVatSlot(watcherVref);
+    assert(virtual, 'promise watcher must be a virtual object');
+    if (watcher.onFulfilled) {
+      assert.typeof(watcher.onFulfilled, 'function');
+    }
+    if (watcher.onRejected) {
+      assert.typeof(watcher.onRejected, 'function');
+    }
```

Comment:
I'm not sure we should care whether those methods exist on the object or not. Maybe just triggering an unhandled rejection at resolution time is enough?

Reply:
If you have multiple promise watchers on the same promise, it's possible this is due to a division of labor (presumably having something to do with how the user code is architected) in which, say, one watcher is dealing with the success case and a different one is dealing with the failure case.  So it seemed reasonable to allow a handler to simply be absent.  In particular, it's only important that _somebody_ handle the rejection as long as somebody actually does.  However, given the consensus that we need to be extra careful about unhandled rejections, I think substituting the default rejection handler (i.e., throw) when `onRejected` is absent is probably the right thing to do, so that swallowing the rejection without action must be a deliberate user choice.  On the other hand, I think treating an absent `onFulfilled` as a no-op seems fine.

Reply:
IMO watchers being an object differ enough from promise reactions as independent callbacks, that requiring a custom watchers (i.e. the ones not using `providePromiseWatcher`) to implement both isn't too much to ask. However I think my point here was that I'm not sure we should perform much validation of the watcher object beside it's an object that can be stored durably.